{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before we begin\n",
    "\n",
    "To talk to a MySQL database, you need to have a mysql \"client\".  **I have already installed this for you!**  But if you ever need to do it yourself, the commands are:\n",
    "\n",
    "\n",
    "        sudo apt-get update\n",
    "        sudo apt-get install mysql-client\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Databases and Structured Query Language (SQL)\n",
    "\n",
    "As Data Scientists, you will frequently want to store data in an organized, structured manner that allows you to do complex queries.  Because you are good Data Scientists, [**you do not use Excel!!!**](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-5-80)\n",
    "\n",
    "In this course, we will only discuss **Relational Databases**, because those are the most common in bioinformatics.  (There are other kinds!!).  So when I say \"database\" I mean \"relational database\".\n",
    "\n",
    "Databases are used to store information in a manner that, when used properly, is:\n",
    "  a) highly structured\n",
    "  b) constrained (i.e. detects errors)\n",
    "  c) transactional (i.e. can undo a command if it discovers a problem)\n",
    "  d) indexed (for speed of search)\n",
    "  e) searchable\n",
    "  \n",
    "The core concept of a database is a **Table**.  Tables contain one particular \"kind\" of information (e.g. a Table could represent a Student, a University, a Book, or a portion of a Clinical Record.\n",
    "\n",
    "Tables contain **Rows** and **Columns** where, generally, every column represents a \"feature\" of that information (e.g. a Student table might have **[\"name\", \"gender\", \"studentID\", \"age\"]** as its columns/features).  Every row represents an \"individual\", and their values for each feature (e.g. a Row in a Student table might have **[\"Mark Wilkinson\", \"M\", \"163483\", \"35\"]** as its values.\n",
    "\n",
    "A Database may have many Tables that represent various kinds of related information.  For example, a library database might have a Books table, a Publishers table, and a Locations table.  A Book has a Publisher, and a Location, so the tables need to be connected to one another.  This is achieved using **keys**.  Generally, every row (individual) in a table has a unique identifier (generally a number), and this is called its **key**.  Because it is unique, it is possible to refer unambiguously to that individual record.\n",
    "\n",
    "I think the easiest way to learn about databases and SQL is to start building one!  We will use the MySQL Docker Container that we created in the previous lesson.  We are going to create a Germplasm database (seed stocks).  It will contain information about the seed (its amount, its harvest date, its location), the germplasm (its species, the allele it carries), and about the genetics related to that allele (the gene_id, the gene name, the protein name, and a link to the GenBank record)\n",
    "\n",
    "(if that Docker Container isn't running, please **docker start course-mysql** now!)\n",
    "\n",
    "**Note:  This Jupyter Notebook is running the Python kernel.  This allows us to use some nice tools in Python (the sql extension and SqlMagic) that provide access to the mysql database server from inside of the Notebook.  You don't need to know any Python to do this.  Note also that you can do exactly the same commands in your Terminal window.**\n",
    "\n",
    "To connect to the MySQL Docker Container from your terminal window, type:\n",
    "\n",
    "     mysql -h 127.0.0.1 -P 3306 --protocol=tcp -u root -p\n",
    " \n",
    "(then enter your password 'root' to access the database)\n",
    " \n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "# SQL\n",
    "\n",
    "Structured Query Language is a way to interact with a database server.  It is used to create, delete, edit, fill, and query tables and their contents.  \n",
    "\n",
    "First, we will learn the SQL commands that allow us to explore the database server, and create new databases and tables..  Later, we will use SQL to put information into those tables.  Finally, we will use SQL to query those tables.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python SQL Extension\n",
    "\n",
    "The commands below are used to connect to the MySQL server in our Docker Container.  You need to execute them ONCE.  In every subsequent Juputer code window, you will have access to the database.\n",
    "\n",
    "all SQL commands are preceded by \n",
    "\n",
    "     %sql \n",
    "     \n",
    "(**only in the Python extension!  Not in your terminal window!**)\n",
    "\n",
    "all SQL commands end with a \";\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "\n",
    "            \n",
    "        #%sql mysql+pymysql://anonymous@ensembldb.ensembl.org/homo_sapiens_core_92_38\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## show databases\n",
    "\n",
    "**show databases** is the command to see what databases exist in the server.  The ones you see now are the default databases that MySQL uses to organize itself.  _**DO NOT TOUCH THESE DATABASES**_  **EVER EVER EVER EVER**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql show databases;\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create database\n",
    "\n",
    "The command to create a database is **create database** (surprise!  ;-) )\n",
    "\n",
    "We will create a database called \"germplasm\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database germplasm;\n",
    "%sql show databases\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## use database_name\n",
    "\n",
    "the **use** command tells the server which database you want to interact with.  Here we will use the database we just created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql use germplasm;\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## show tables\n",
    "\n",
    "The show tables command shows what tables the database contains (right now, none!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Planning your data structure\n",
    "\n",
    "This is the hard part.  What does our data \"look like\" in a well-structured, relational format?\n",
    "\n",
    "Starting simply:\n",
    "\n",
    "<center>stock table</center>\n",
    "\n",
    "  amount  |  date  |  location  \n",
    " --- | --- | --- \n",
    "  5  | 10/5/2013 |  Room 2234  \n",
    "  9.8  | 12/1/2015 |  Room 998  \n",
    "\n",
    "\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "<center>germplasm table</center>\n",
    "\n",
    "  taxonid  |  allele\n",
    " --- | --- \n",
    "  4150  | def-1\n",
    "  3701  | ap3\n",
    "  \n",
    "--------------------------------\n",
    "\n",
    "<center>gene table</center>\n",
    "\n",
    "  gene  |  gene_name  |  embl\n",
    " --- | ---  | --- \n",
    "  DEF  | Deficiens  | https://www.ebi.ac.uk/ena/data/view/AB516402\n",
    "  AP3  | Apetala3   |   https://www.ebi.ac.uk/ena/data/view/AF056541\n",
    "  \n",
    "  \n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add indexes\n",
    "\n",
    "It is usually a good idea to have an index column on every table, so let's add that first:\n",
    "\n",
    "\n",
    "<center>stock table</center>\n",
    "\n",
    "id  |  amount  |  date  |  location  \n",
    "---  | --- | --- | --- \n",
    "1   |  5  | 10/5/2013 |  Room 2234  \n",
    "2   |   9.8  | 12/1/2015 |  Room 998  \n",
    "\n",
    "\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "<center>germplasm table</center>\n",
    "\n",
    "id  |  taxonid  |  allele\n",
    "--- |  --- | --- \n",
    "1  |  4150  | def-1\n",
    "2  |   3701  | ap3\n",
    "  \n",
    "--------------------------------\n",
    "\n",
    "<center>gene table</center>\n",
    "\n",
    "id  |  gene  |  gene_name  |  embl\n",
    "--- | --- | ---  | --- \n",
    "1  |  DEF  | Deficiens  | https://www.ebi.ac.uk/ena/data/view/AB516402\n",
    "2  |  AP3  | Apetala3   |   https://www.ebi.ac.uk/ena/data/view/AF056541\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  find linkages\n",
    "\n",
    "* Every germplasm has a stock record.  This is a 1:1 relationship.\n",
    "* Every germplasm represents a specific gene.  This is a 1:1 relationship\n",
    "\n",
    "So every germplasm must point to the index of a stock, and also to the index of a gene\n",
    "\n",
    "Adding that into our tables we have:\n",
    "\n",
    "\n",
    "\n",
    "<center>stock table</center>\n",
    "\n",
    "id  |  amount  |  date  |  location  \n",
    "---  | --- | --- | --- \n",
    "1   |  5  | 10/5/2013 |  Room 2234  \n",
    "2   |   9.8  | 12/1/2015 |  Room 998  \n",
    "\n",
    "\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "<center>germplasm table</center>\n",
    "\n",
    "id  |  taxonid  |  allele  |  stock_id  |  genetics_id\n",
    "--- |  --- | ---  | --- | ---\n",
    "1  |  4150  | def-1  | 2   |  1\n",
    "2  |   3701  | ap3   | 1   |  2\n",
    "  \n",
    "--------------------------------\n",
    "\n",
    "<center>gene table</center>\n",
    "\n",
    "id  |  gene  |  gene_name  |  embl\n",
    "--- | --- | ---  | --- \n",
    "1  |  DEF  | Deficiens  | https://www.ebi.ac.uk/ena/data/view/AB516402\n",
    "2  |  AP3  | Apetala3   |   https://www.ebi.ac.uk/ena/data/view/AF056541\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## data types in MySQL\n",
    "\n",
    "I will not discuss [all MySQL Datatypes](https://dev.mysql.com/doc/refman/5.7/en/data-types.html), but we will look at only the ones we need.  We need:\n",
    "\n",
    "* Integers (type INTEGER)\n",
    "* Floating point (type FLOAT)\n",
    "* Date  (type DATE [in yyyy-mm-dd format](https://dev.mysql.com/doc/refman/5.7/en/datetime.html) )\n",
    "* Characters (small, variable-length --> type [VARCHAR(x)](https://dev.mysql.com/doc/refman/5.7/en/char.html) )\n",
    "\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "## create table \n",
    "\n",
    "tables are created using the **create table** command (surprise!)\n",
    "\n",
    "The [syntax of create table](https://dev.mysql.com/doc/refman/5.7/en/create-table.html) can be quite complicated, but we are only going to do the most simple examples.\n",
    "\n",
    "    create table table_name (column_name column_definition, column_name column_definition, ........)\n",
    "    \n",
    "column definitions include the data-type, and other options like if it is allowed to be null(blank), or if it should be treated as an \"index\" column.\n",
    "\n",
    "Examples are easier to understand than words... so here are our table definitions:\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql drop table stock\n",
    "%sql CREATE TABLE stock(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, amount FLOAT NOT NULL, date DATE NOT NULL, location VARCHAR(20) NOT NULL);\n",
    "%sql DESCRIBE stock\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql drop table germplasm\n",
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, taxonid INTEGER NOT NULL, allele VARCHAR(10) NOT NULL, stock_id INTEGER NOT NULL, gene_id INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql drop table gene\n",
    "%sql CREATE TABLE gene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, gene VARCHAR(10) NOT NULL, gene_name VARCHAR(30) NOT NULL, embl VARCHAR(70) NOT NULL);\n",
    "%sql DESCRIBE gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## loading data\n",
    "\n",
    "There are many ways to import data into MySQL.  If you have data in another (identical) MySQL database, you can \"dump\" the data, and then import it directly.  If you have tab or comma-delimited (tsv, csv) you can **sometimes** import it directly from these formats.  You can also enter data using SQL itself.  This is usually the safest way, when you have to keep multiple tables synchronized (as we do, since the germplasm table is \"linked to\" the other two tables)\n",
    "\n",
    "## insert into\n",
    "\n",
    "The command to load data is:\n",
    "\n",
    "    insert into table_name (field1, field2, field3) values (value1, value2, value3)\n",
    "    \n",
    "Now... what data do we need to add, in what order?\n",
    "\n",
    "The germplasm table needs the ID number from both the gene table and the stock table, so we cannot enter the germplasm information first.  We must therefore enter the gene and stock data first.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE - we DO NOT put data into the \"id\" column!  This column is auto_increment, so it \"magically\" creates its own value\n",
    "%sql INSERT INTO gene (gene, gene_name, embl) VALUES ('DEF', \"Deficiens\", 'https://www.ebi.ac.uk/ena/data/view/AB516402');\n",
    "%sql INSERT INTO gene (gene, gene_name, embl) VALUES ('AP3', \"Apetala3\", 'https://www.ebi.ac.uk/ena/data/view/AF056541');\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT last_insert_id();  # just to show you that this function exists!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql INSERT INTO stock(amount, date, location) VALUES (5, '2013-05-10', 'Room 2234');\n",
    "%sql INSERT INTO stock(amount, date, location) VALUES (9.8, '2015-01-12', 'Room 998');\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Almost ready!  \n",
    "\n",
    "We now need to know the index numbers from the stock and gene databases that correspond to the data for the germplasm table.  For this, we need to learn another function:  **select**\n",
    "\n",
    "## Select statements\n",
    "\n",
    "**Select** is the command used to query the database.  We will look in more detail later, but now all you need to know is that the most basic structure is:\n",
    "\n",
    "     select * from table_name\n",
    "     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM stock;  # notice that the id number was automatically generated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "Just a reminder, our germplasm data is:\n",
    "\n",
    "id  | \ttaxonid  |  allele  | \tstock_id  |\tgene_id\n",
    "--- | --- | --- | --- | --- |\n",
    "1 \t | 4150 \t|def-1 \t\n",
    "2 \t| 3701 |\tap3 \t\n",
    "\n",
    "\n",
    "We need to connect the *germplasm* table **gene_id** to the appropriate **id** from the *gene* table.  i.e.\n",
    "\n",
    "    def-1 allele  --->  DEF gene  (id = 1)\n",
    "    ap3 allele  --->  AP3 gene  (id = 2)\n",
    "   \n",
    "We need to connect the *germplasm* table **stock_id** to the appropriate **id** from the *stock* table.  i.e.\n",
    "\n",
    "    def-1 allele  --->  Room 998 (id = 2)\n",
    "    ap3 allele  --->  Room 2234  (id = 1)\n",
    "\n",
    "Now we are ready to do our (\"manual\") insert of data into the *germplasm* table:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql INSERT INTO germplasm (taxonid, allele, stock_id, gene_id) VALUES (4150, 'def-1', 2, 1 );\n",
    "%sql INSERT INTO germplasm (taxonid, allele, stock_id, gene_id) VALUES (3701, 'ap3', 1, 2 );\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## SQL UPDATE  &  SQL WHERE\n",
    "\n",
    "Imagine that we are going to plant some seed from our def-1 germplasm.  We need to update the *stock* record to show that there is now less seed available.  We do this using an [UPDATE statement](https://www.techonthenet.com/mysql/update.php).  UPDATE is used to change the values of a particular column or set of columns.  But we don't want to change **all** of the values in that column, we only want to change the values for the DEF stock.  For that, we need a WHERE clause.  \n",
    "\n",
    "WHERE allows you to set the conditions for an update.  The general form is:\n",
    "\n",
    "     UPDATE table_name SET column = value WHERE column = value;\n",
    "     \n",
    "We will sow 1g of seed from DEF  (stock.id = 2)   (note that I am now starting to use the MySQL syntax for referring to *table*.**column** - the tablename followed by a \".\" followed by the column name).\n",
    "\n",
    "The simplest UPDATE statement is:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql UPDATE stock SET amount = 8.8 WHERE id = 2;\n",
    "%sql SELECT * FROM stock;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "This simple solution is not very \"friendly\"... you are asking the database user to already know what the remaining amount is!  It would be better if we simply reduced the amount by 1g.  \n",
    "\n",
    "That is done using in-line equations, like this:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql UPDATE stock SET amount = amount-1 WHERE id = 2;\n",
    "%sql SELECT * FROM stock;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "## Using indexes and 'joining' tables\n",
    "\n",
    "The UPDATE we did is still not very friendly!  My stock table does not have any information about what gene or allele is in that stock, so we have to **know** that the stock record is stock.id=2.  This is bad!\n",
    "\n",
    "It would be better if we could say \"plant 1 gram of the stock that represents gene record DEF\", but that information exists in two different tables.  How do we join tables?\n",
    "\n",
    "This is the main purpose of the \"id\" column.  Note that, when we defined that column, we said that it is \"auto_increment, not null, primary key\", meaning that every record must have an id, and every id must be unique (NOTE: an auto-increment id *should never be manually modified/added*!!!  <span style=\"color:red;\">You Have Been Warned!!!</span>).  Being a 'primary key' means that this column was intended to be the \"pointer\" from other tables in the database (like our germplasm table, that points to the id of the stock, and the id of the gene, tables)\n",
    "\n",
    "When using UPDATE with multiple tables, we must name all of the tables, and then make the connection between them in the \"where\" clause, using *table*.**column** notation.  \n",
    "\n",
    "The update clause below shows how this is done (a \"\\\\\" character means that the command continues on the next line):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql UPDATE stock, germplasm SET stock.amount = stock.amount-1 \\\n",
    "WHERE \\\n",
    "stock.id = germplasm.stock_id \\\n",
    "AND \\\n",
    "germplasm.allele = 'def-1';              \n",
    " \n",
    "%sql SELECT * FROM stock;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "# Challenges for you!\n",
    "\n",
    "1. (hard) when we plant our seeds, we should update both the quantity, and the date.  What does that UPDATE statement look like?\n",
    "\n",
    "2. (very hard!) when we plant our seed, instead of using the allele designation (def-1) I want to use the gene designation (DEF).  This query spans **all three tables**.  What does the UPDATE statement look like?\n",
    "\n",
    "<span style=\"visibility:hidden;\">\n",
    "    Challenge 1\n",
    "    %sql UPDATE stock,germplasm SET stock.amount=stock.amount-1, stock.date=\"2018-09-06\" WHERE \\\n",
    "     stock.id = germplasm.stock_id AND \\\n",
    "     germplasm.allele='def-1';\n",
    "    Challenge2\n",
    "%sql UPDATE stock,germplasm,gene SET stock.amount=stock.amount-0.2, stock.date=\"2018-09-06\" WHERE \\\n",
    "     stock.id = germplasm.stock_id AND \\\n",
    "     gene.id = germplasm.gene_id AND \\\n",
    "     gene.gene='DEF';   \n",
    "</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# challenge 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# challenge 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "# SELECT queries\n",
    "\n",
    "Querying the data is the most common operation on a database.  You have seen simple SELECT queries, but now we will look at more complex ones.\n",
    "\n",
    "The general structure is:\n",
    "\n",
    "     SELECT table1.column1, ... FROM table1, ... WHERE condition1 [AND|OR] condition2....\n",
    "     \n",
    "You probably understand this enough to show you the query that will show you all of the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene, stock, germplasm WHERE \\\n",
    "     germplasm.stock_id = stock.id AND \\\n",
    "     germplasm.gene_id = gene.id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Dealing with missing records - JOIN clauses\n",
    "\n",
    "**Credit for the Venn diagrams used in this section goes to [Flatiron School](https://learn.co/) and are linked from their tutorial on [JOINs in SQL](https://learn.co/lessons/sql-complex-joins-readme) published under the [CC-BY-NC 4.0 license](https://creativecommons.org/licenses/by-nc-sa/4.0/)**\n",
    "\n",
    "Your first database will probably be complete, and perfect!  You will be very proud of it! ;-)\n",
    "\n",
    "Over time, things will happen.  Records will be deleted, and records will be added where there is incomplete information, for example, a germplasm stock record where the gene is unknown.  You should think about these situations, because there are NO RULES telling you what you should do!  You have to make a decision - a *policy* for your database - and you should follow that policy like a religion!\n",
    "\n",
    "For example:\n",
    "\n",
    "*  If there is no known gene for a given germplasm, what does in the stock.allele column?  What goes in the stock.gene_id column?  What goes in the gene table?  Discuss....\n",
    "*  If a stock is fully planted - no more seeds - what do you do?  Should you delete the record?  If you do, then your germplasm table is linked through the stock_id to a stock that no longer exists.  If you gather more seed in the future, is that the same stock?  (answer:  NO, it is not!!!).... so do you update the existing stock record to say there is now 10g of seed?  \n",
    "*  Remember, you are trapped!  In your table definition you declared all columns to be \"NOT NULL\", meaning that if the row exists, there must be a value for each column in the row!  What do you do if there isn't a value to put into that column?  \n",
    "  *  zero?  \n",
    "  *  What if you change the column definition to allow NULL?\n",
    "  *  What does NULL mean?  What does zero mean?  \n",
    "  *  How does software respond to NULL or zero values?  (you don't know this yet, but we can talk about it)\n",
    "\n",
    "\n",
    "For our database, I am going to suggest this policy:\n",
    "1. if we don't know the allele, we put \"unknown\" in the allele column\n",
    "2. We put '0' into the gene_id column (auto_increment starts with 1 in the gene table, so a 0 will match nothing!)  \n",
    "3. We DO NOT add a gene record at all.\n",
    "\n",
    "Let's add a record like this one to our database:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql INSERT INTO stock(amount, date, location) VALUES (23, '2018-05-12', 'Room 289');\n",
    "%sql INSERT INTO germplasm (taxonid, allele, stock_id, gene_id) VALUES (4150, 'unknown', LAST_INSERT_ID(), 0 );\n",
    "# note that I am using LAST_INSERT_ID to capture the auto_increment value from the stock table insert\n",
    "# this ensures that the germplasm and stock tables are 'synchronized'\n",
    "%sql SELECT * FROM germplasm;\n",
    "#%sql SELECT * FROM gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "That looks good!  ...but we have just created a problem!  gene_id=0 doesn't exist in the gene table, so what happens with our beautiful SELECT query that we just created above?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene, germplasm WHERE \\\n",
    "     germplasm.gene_id = gene.id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### OH CRAP!!!!  We lost our data!!\n",
    "\n",
    "Our \"unknown\" germplasm has disappeared!!  Or has it?\n",
    "\n",
    "The problem is that stock.gene_id = gene.id failed for the \"unknown\" record, and so it isn't reflected in the output from the query.  THIS IS BAD, if (for example) you were trying to take an inventory of all germplasm stocks you had!  \n",
    "\n",
    "How do we solve this?  The answer is to use SQL's \"JOIN\" instruction.\n",
    "\n",
    "There are four kinds of JOIN:  INNER, LEFT OUTER, RIGHT OUTER, and FULL OUTER.\n",
    "\n",
    "The join we are doing with our current SELECT query is an INNER join.  Using a Venn diagram, the query looks like this:\n",
    "\n",
    "<a href='https://learn.co/lessons/sql-complex-joins-readme'><img src='http://readme-pics.s3.amazonaws.com/Inner%20Join%20Venn%20Diagram.png' width=300px/></a>\n",
    "\n",
    "\n",
    "Effectively, the intersection where BOTH the 'left' (gene.id) and 'right' (germplasm.id) are true.\n",
    "\n",
    "You can duplicate this behavior using the INNER JOIN instruction.  The syntax is a little bit different - look:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene INNER JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "What we want is a query that allows one side to be \"missing/absent/NULL\", but the other side to exist.\n",
    "\n",
    "Perhaps we need a \"LEFT JOIN\"?\n",
    "\n",
    "    ... gene LEFT JOIN germplasm ...\n",
    "\n",
    "\n",
    "Again, in this situation, \"LEFT\" means the table on the Left side of the SQL JOIN statement (gene)\n",
    "\n",
    "As a Venn diagram, Left joins look like this (ignore the :\n",
    "\n",
    "<a href='https://learn.co/lessons/sql-complex-joins-readme'><img src='http://readme-pics.s3.amazonaws.com/Left%20Outer%20Join%20Venn%20Diagram.png' width=300px/></a>\n",
    "\n",
    "What it means is that, **in addition to the perfect matches at the intersection**, the record on the left (the gene record) should be included in the result set, **even if it doesn't match** with a germplasm record (on the right).  Is that the solution to our problem?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene LEFT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PFFFFFF!!   No, that was not the solution\n",
    "\n",
    "Why?\n",
    "\n",
    "What about a RIGHT JOIN?\n",
    "\n",
    "    ... gene RIGHT JOIN germplasm ...\n",
    "\n",
    "<a href='https://learn.co/lessons/sql-complex-joins-readme'><img src='http://readme-pics.s3.amazonaws.com/Right%20Outer%20Join%20Venn%20Diagram.png' width=300px/></a>\n",
    "\n",
    "\n",
    "\n",
    "Again, in this situation, \"RIGHT\" means the table on the Right side of the SQL JOIN statement (germplasm).  So the germplasm record should be included in the result set, even if a gene record does not exist.  ...that sounds much more likely to be correct!  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene RIGHT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Voila!!\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Your turn\n",
    "\n",
    "1)  Create another record, where in this case, there is no **stock**, but there is a germplasm and a gene record.  \n",
    "2)  Create the JOIN query between germplasm and stock that includes all germplasm records\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "# Other SELECT \"magic\"\n",
    "\n",
    "You can do many other useful things with SELECTS, such as:\n",
    "\n",
    "## COUNT()\n",
    "\n",
    "If you want to count the number of records returned from a query, use the **COUNT() AS your_name** function:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number Of Matches\" FROM gene RIGHT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SUM(), AVG(), MAX()\n",
    "\n",
    "You can do mathematical functions on results also, for example, you can take the SUM of a column - how much seed do we have in total?  \n",
    "\n",
    "(look carefully at this query!  It's quite complicated!):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT SUM(stock.amount) FROM gene RIGHT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id \\\n",
    "     INNER JOIN stock ON germplasm.stock_id = stock.id;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "Or you could take the **average AVG()** of a column - what is the average quantity of seed we have?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT AVG(stock.amount) FROM gene RIGHT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id \\\n",
    "     INNER JOIN stock ON germplasm.stock_id = stock.id;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "Or you could take the **max MAX()** value of a column - what is the largest quantity of seed we have in our stocks?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT MAX(stock.amount) FROM gene RIGHT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id \\\n",
    "     INNER JOIN stock ON germplasm.stock_id = stock.id;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ORDER BY\n",
    "\n",
    "You can put your results in a specific order:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT gene.gene_name, stock.amount FROM gene RIGHT JOIN germplasm ON \\\n",
    "     germplasm.gene_id = gene.id \\\n",
    "     INNER JOIN stock ON germplasm.stock_id = stock.id \\\n",
    "     ORDER BY stock.amount DESC;  # change this to ASC\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "## Conclusion\n",
    "\n",
    "1) Databases are a very powerful way to store structured information - far far better than Excel Spreadsheets!  \n",
    "2) It will take you years (literally, years!) to become an expert in MySQL!  We have only explored the most common functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql drop database germplasm;"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
